### PR TITLE
Optimize burden estimate upload

### PR DESCRIPTION
**Pull request by MartinEden**
The continuing adventures of https://vimc.myjetbrains.com/youtrack/issue/VIMC-781

### CODE REVIEW

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt [hunk 1]
```diff
@@ -86,40 +89,84 @@ class JooqBurdenEstimateRepository(
...
+            .map { it[COUNTRY.ID] }
+            .toHashSet()
+
     private fun newBurdenEstimateRecord(
             setId: Int,
             estimate: BurdenEstimate,
             outcomeId: Int,
             outcomeValue: BigDecimal?
-    ): BurdenEstimateRecord
+    ): List<Any?>
```
Comment (littledice) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: can we name this function in a way that reflects the new output type? like `listOfBurdenEstimateValues` (or something better)?

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt [hunk 2]
```diff
@@ -86,40 +89,84 @@ class JooqBurdenEstimateRepository(
...
+                        if (estimate.disease != expectedDisease)
+                        {
+                            throw InconsistentDataError("Provided estimate lists disease as '${estimate.disease}' but scenario is for disease '$expectedDisease'")
+                        }
+                        if (estimate.country !in countries)
+                        {
+                            throw UnknownObjectError(estimate.country, "country")
+                        }
+
+                        writer.writeRecord(newBurdenEstimateRecord(setId, estimate, cohortSizeId, estimate.cohortSize))
```
Comment (littledice) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: `writeRecord` is a little ambiguous; this is writing a line to a text file right? maybe `writeLine` or `writeValues` or something similar?

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt [hunk 3]
```diff
@@ -2,11 +2,61 @@ package org.vaccineimpact.api.db
...
+fun CopyManager.copyInto(table: TableImpl<*>, stream: InputStream, fields: List<Field<*>>)
+{
+    val fieldList = fields.joinToString(", ") { it.name }
+    val query = "COPY ${table.name} ($fieldList) FROM STDIN"
+    this.copyIn(query, stream)
+}
+
+fun DSLContext.withoutCheckingForeignKeyConstraints(table: TableImpl<*>, work: () -> Unit)
+{
+    val triggers = this.getTriggersFor(table)
```
Comment (littledice) [inline src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: I'm sure its' obvious that it is working from the performance gain but would be nice to test this.

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    constructor(stream: OutputStream)
+        : this(stream.writer())
+
+    // From the documentation: "In particular, the following characters must be preceded by a backslash
+    // if they appear as part of a column value: backslash itself, newline, carriage return, and the current
+    // delimiter character."
+    private val illegalCharacters = Regex("\\|\n|\r|\t")
+
+    fun writeRecord(values: Iterable<Any?>)
+    {
```
Comment (littledice) [inline src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt]: Can we have some unit tests for the escaping logic?

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt [hunk 5]
```diff
@@ -2,11 +2,61 @@ package org.vaccineimpact.api.db
...
+}
+
+private fun DSLContext.setTriggerEnabled(table: TableImpl<*>, triggerName: String, action: Toggle)
+{
+    val actionText = when (action)
+    {
+        Toggle.Enable -> "ENABLE"
+        Toggle.Disable -> "DISABLE"
+    }
+    this.execute("""ALTER TABLE ${table.name} $actionText TRIGGER "$triggerName"""")
```
Comment (littledice) [inline src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: Validation higher up the call chain should prevent this happening, but what happens if somehow, values are copied to a table that don't obey constraints? Is the resulting error message relatively clear/descriptive? Would be good to test this.

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt [hunk 6]
```diff
@@ -86,40 +89,84 @@ class JooqBurdenEstimateRepository(
...
+                PostgresCopyWriter(stream).use { writer ->
+                    for (estimate in estimates)
+                    {
+                        if (estimate.disease != expectedDisease)
+                        {
+                            throw InconsistentDataError("Provided estimate lists disease as '${estimate.disease}' but scenario is for disease '$expectedDisease'")
+                        }
+                        if (estimate.country !in countries)
+                        {
+                            throw UnknownObjectError(estimate.country, "country")
```
Comment (littledice) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: There's no test specifically for country validation, lets add one (relates to the comment on `JooqExtensions` about what happens if bad values are added)

**Comment by littledice** [comment]
Looks good! Mainly just requests for more tests.

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt (hunk 3 above)
Comment (MartinEden) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: Sorry, not sure what you mean. Do you mean you want a test that disabling the triggers is faster?

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt (hunk 5 above)
Comment (MartinEden) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: There won't be any error message. If, while the foreign key constraints are disabled, we add data that disobeys the foreign key constraints, the data will just become inconsistent, and the next time someone does a join on that table we will get missing rows. That's why it's important that we don't insert any inconsistent data!

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt (hunk 3 above)
Comment (littledice) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: I meant a test to check that the triggers really do get disabled

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt (hunk 5 above)
Comment (littledice) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: Ok, makes sense

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt (hunk 4 above)
Comment (MartinEden) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt]: Done, and they caught a bug, so thanks for catching this!

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt (hunk 2 above)
Comment (MartinEden) [reply src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: It is writing to a text file, but each line of the text file still corresponds to inserting a row into the database, so I felt like here (and with your comment below) I didn't want to go as vague as "values". It's still a record, although it's not a Jooq record. Maybe `writeRow` and `newBurdenEstimateRow`?

Comment (littledice) [reply src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: Those sound good to me :+1:

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt (hunk 4 above)
Comment (littledice) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/PostgresCopyWriter.kt]: Nice!

File: src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt (hunk 3 above)
Comment (MartinEden) [reply src/databaseInterface/src/main/kotlin/org/vaccineimpact/api/db/JooqExtensions.kt]: Added!

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt (hunk 2 above)
Comment (MartinEden) [reply src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqBurdenEstimateRepository.kt]: Made the changes.

**Pull request merged**